api/types/second_factor.go

```diff
@@ -0,0 +1,172 @@
+		return constants.SecondFactorOTP
+	default:
+		return constants.SecondFactorOff
+	}
+}
+
+// secondFactorsFromLegacySecondFactor returns the list of SecondFactorTypes supported by the given second factor type.
+func secondFactorsFromLegacySecondFactor(sf constants.SecondFactorType) []SecondFactorType {
+	switch sf {
+	case constants.SecondFactorOff:
+		return nil
+	case constants.SecondFactorOptional, constants.SecondFactorOn:
+		return []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN, SecondFactorType_SECOND_FACTOR_TYPE_OTP}
+	case constants.SecondFactorOTP:
+		return []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_OTP}
+	case constants.SecondFactorWebauthn:
+		return []SecondFactorType{SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN}
+	default:
+		return nil
+	}
+}
+
+// MarshalJSON marshals SecondFactorType to string.
+func (s *SecondFactorType) MarshalYAML() (interface{}, error) {
+	val, err := s.encode()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return val, nil
+}
+
+// UnmarshalYAML supports parsing SecondFactorType from string.
+func (s *SecondFactorType) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var val interface{}
+	err := unmarshal(&val)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = s.decode(val)
+	return trace.Wrap(err)
+}
+
+// MarshalJSON marshals SecondFactorType to string.
+func (s *SecondFactorType) MarshalJSON() ([]byte, error) {
+	val, err := s.encode()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	out, err := json.Marshal(val)
+	return out, trace.Wrap(err)
+}
+
+// UnmarshalJSON supports parsing SecondFactorType from string.
+func (s *SecondFactorType) UnmarshalJSON(data []byte) error {
+	var val interface{}
+	err := json.Unmarshal(data, &val)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = s.decode(val)
+	return trace.Wrap(err)
+}
+
+const (
+	// secondFactorTypeOTPString is the string representation of SecondFactorType_SECOND_FACTOR_TYPE_OTP
+	secondFactorTypeOTPString = "otp"
+	// secondFactorTypeWebauthnString is the string representation of SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN
+	secondFactorTypeWebauthnString = "webauthn"
+	// secondFactorTypeSSOString is the string representation of SecondFactorType_SECOND_FACTOR_TYPE_SSO
+	secondFactorTypeSSOString = "sso"
+)
+
+// ToString returns the user friendly string representation of the second factor type.
+func (s *SecondFactorType) ToString() string {
+	str, _ := s.encode()
+	return str
+}
+
+func (s *SecondFactorType) encode() (string, error) {
+	switch *s {
+	case SecondFactorType_SECOND_FACTOR_TYPE_UNSPECIFIED:
+		return "", nil
+	case SecondFactorType_SECOND_FACTOR_TYPE_OTP:
+		return secondFactorTypeOTPString, nil
+	case SecondFactorType_SECOND_FACTOR_TYPE_WEBAUTHN:
+		return secondFactorTypeWebauthnString, nil
+	case SecondFactorType_SECOND_FACTOR_TYPE_SSO:
+		return secondFactorTypeSSOString, nil
+	default:
+		return "", trace.BadParameter("SecondFactorType invalid value %v", *s)
```

Comment:
```suggestion
		return "", trace.BadParameter("invalid SecondFactorType value: %q", *s)
```

So we avoid capitalization at the start. Same for others.

- https://go.dev/wiki/CodeReviewComments#error-strings